Add EVP functionality to create domain params and keys by user data

levitte opened the pull request:
This is the EVP operation that corresponds to creating direct RSA, DH
and DSA keys and set their numbers, to then assign them to an EVP_PKEY,
but done entirely using an algorithm agnostic EVP interface.

- [x] Documentation
- [x] Test

levitte committed:
Add EVP functionality to create domain params and keys by user data

This is the EVP operation that corresponds to creating direct RSA, DH
and DSA keys and set their numbers, to then assign them to an EVP_PKEY,
but done entirely using an algorithm agnostic EVP interface.

levitte commented:
@slontis, I think you're interested in this

levitte commented:
Note that the added functionality deviates slightly from a previous pattern.  The param and key *generation* functionality has these functions:

``` C
int EVP_PKEY_paramgen_init(EVP_PKEY_CTX *ctx);
int EVP_PKEY_paramgen(EVP_PKEY_CTX *ctx, EVP_PKEY **ppkey);
int EVP_PKEY_keygen_init(EVP_PKEY_CTX *ctx);
int EVP_PKEY_keygen(EVP_PKEY_CTX *ctx, EVP_PKEY **ppkey);
```

The making functions are these:

``` C
int EVP_PKEY_parammake_init(EVP_PKEY_CTX *ctx);
int EVP_PKEY_keymake_init(EVP_PKEY_CTX *ctx);
int EVP_PKEY_make(EVP_PKEY_CTX *ctx, EVP_PKEY **ppkey, OSSL_PARAM param[]);
```

Two functions to init, but just one to actually make.  I find it quite unnecessary to have two making functions when they can differentiate by the operation number that's already assigned in the EVP_PKEY context.

levitte commented:
This will remain WIP until #10169 and #10184 are in

slontis commented on CHANGES:
```diff
@@ -9,6 +9,14 @@
 
  Changes between 1.1.1 and 3.0.0 [xx XXX xxxx]
 
+  *) Added functionality to create an EVP_PKEY context based on data
```
This seems like a confusing description here. The API docs are clearer than this.

levitte replied:
Uhmmmmm...  this description has nothing to do with this PR, it belongs with #10184...  did I mix the CHANGES changes up? ...  yup, this seems to be a rebasse goof-up

levitte replied:
Fixed

levitte commented on crypto/evp/keymgmt_lib.c:
```diff
@@ -200,6 +200,28 @@ void evp_keymgmt_clear_pkey_cache(EVP_PKEY *pk)
+#else
+        target->pkeys[0].provdata = provdata;
+        target->pkeys[0].domainparams = domainparams;
+#endif
```
As soon as #10169 is in, lines 215-220 will be reduced to this:

``` C
        target->pkeys[0].provdata = provdata;
        target->pkeys[0].domainparams = domainparams;
```

levitte replied:
Done

levitte commented:
Now that #10169 and #10184 have been merged, I rebased this.  It's still in draft, as a test program should be written for this.

levitte committed:
Add a test for EVP_PKEY_keymake() and EVP_PKEY_make()

This test is a bit lame, but will either be completed as functionality
is added in the default provider, or the new functions may start
getting used in evp_test.c and this program will disappear.

levitte commented:
Not WIP any more

slontis commented:
I am not completely sold on the name 'make'.
EVP_PKEY_fromparams
EVP_PKEY_fromraw

which could then match something like..
EVP_PKEY_toparams

levitte commented:
I have no problems changing to those names.  But...  `EVP_PKEY_fromraw`?  Does that imply that we need that "raw key" functionality in providers?  I thought that was a hack needed to the MAC-as-PKEY stuff, which is entirely libcrypto side...  or is this somethiing else?

levitte commented:
Although...  thinking a bit more about it, those names may add confusion.  There are two thing that we need to build, domain parameters (classic *p* and *g* and possibly *q* for DH / DSA), and the key itself...  so what would we then call the init functions?  `EVP_fromparams_params_init`, `EVP_fromparams_key_init`?

My "I have no problems..." was a bit premature, it seems

slontis commented:
The fromraw was just an alternate suggestion for the name.. I think the fromparams is better.
I was hoping I could use these for the self tests (which sit in the fips provider).

slontis commented:
We have given params 2 different meanings now, which makes using params very confusing.
Coming up with names is hard :)
These functions load a PKEY with custom user data. So something that indicates we are setting userdata into a key would be a good name.

EVP_PKEY_param_setuserdata_init
EVP_PKEY_key_setuserdata_init
EVP_PKEY_setuserdata

levitte commented:
Hmmm, I think we're getting somewhere...  Mind you, I *like* having "from" in there, so merging all ideas together, what would you say about this?

EVP_PKEY_param_fromdata_init
EVP_PKEY_key_fromdata_init
EVP_PKEY_fromdata

(could be "fromuserdata" too, but I'm unsure if "user" is really needed in there)

slontis commented:
Yeah sounds good to me - I like 'from' because you can then do 'to' for the other direction if required.

levitte commented:
I put this back as WIP 'cause I realised it's not quite ready yet.

So here's an auxiliary function that I forgot to implement; the function that returns a `const OSSL_PARAM` array describing what can be used with (presumed) `EVP_PKEY_fromdata`.  Is `EVP_PKEY_fromdata_settable` good enough?  It's not *exactly* wrong, but not perfect (enough) either...

slontis commented:
> In my mind, it's not _exactly_ wrong, but not perfect (enough) either...

I guess that matches the other ones.. 
Does this return every type of key data (for RSA, DSA, DH, EC).

levitte commented:
It should return the data corresponding to the algorithm, so should take a EVP_PKEY_CTX as argument.

slontis commented:
Which comes from EVP_PKEY_CTX_new_provided - which has the type.. Ok.

levitte commented:
(or even the old ones, 'cause they do set the algorithm name too, just not the props)

levitte commented:
Ok, I'll got with that.  It's also up for reconsideration as long as we haven't released 3.0, so I'm not terribly worried about getting it absolutely right in its final form now, as long ass it's good enough for us to be happy.

levitte commented:
I'm wondering about the `settable` names again.  Is there a reason for them to be tied by name to some specific operation?  Can't they just be this `EVP_PKEY_param_settable()` and `EVP_PKEY_key_settable`?

levitte commented:
Ping.  Question needs answering

beldmit commented:
> The making functions are these:
> 
> ```c
> int EVP_PKEY_parammake_init(EVP_PKEY_CTX *ctx);
> int EVP_PKEY_keymake_init(EVP_PKEY_CTX *ctx);
> int EVP_PKEY_make(EVP_PKEY_CTX *ctx, EVP_PKEY **ppkey, OSSL_PARAM param[]);
> ```
> 
> Two functions to init, but just one to actually make. I find it quite unnecessary to have two making functions when they can differentiate by the operation number that's already assigned in the EVP_PKEY context.

Is it reasonable for the algorithms with parameters treated as special objects?

levitte commented:
> Is it reasonable for the algorithms with parameters treated as special objects?

Not really. A key is a key is a key. That it carries domain parameters should not make a difference.

beldmit commented:
> > Is it reasonable for the algorithms with parameters treated as special objects?
> 
> Not really. A key is a key is a key. That it carries domain parameters should not make a difference.

It may take time to generate parameters for a, e.g., DSA key, and one may want possible to manipulate the parameters generation and the key generation separately, and/or serialize the params. If it is out of the scope of the API you suggest, it's OK.

levitte commented:
> It may take time to generate parameters for a, e.g., DSA key, and one may want possible to manipulate the parameters generation and the key generation separately, and/or serialize the params. If it is out of the scope of the API you suggest, it's OK.

Not sure I understand...  You *do* generate the parameters separately, and they end up in a `EVP_PKEY`, and then you *do* generate the key itself separately, and it ends up in another `EVP_PKEY`.

levitte commented:
However, please note that *this PR isn't about domparam or key generation*.  There will be another PR to do that.

levitte commented:
The question I'm still waiting for an answer to is this:

> I'm wondering about the `settable` names again.  Is there a reason for them to be tied by name to some specific operation?  Can't they just be this `EVP_PKEY_param_settable()` and `EVP_PKEY_key_settable`?

slontis commented:
> I'm wondering about the `settable` names again

If this operation is only related to the  fromdata  methods then the name is fine..

levitte commented:
OK, I will stay with that, but will add a TODO comment to re-evaluate the name later.

levitte commented:
This is now ready for final evaluation

slontis commented on CHANGES:
```diff
@@ -9,6 +9,14 @@
 
+  *) Added functionality to create an EVP_PKEY from user data.  This
+     is effectively the same as creating an RSA, DH or DSA object and
+     assign them numbers and then assign them to an EVP_PKEY, but is
```
Although I know what you mean - it is not very clear. Maybe somnething like...

This is equivalent to creating a EVP_PKEY using RSA, DH, or DSA methods.

levitte replied:
How about I simply remove "and assign them numbers"?

slontis commented on doc/man3/EVP_PKEY_fromdata.pod:
```diff
@@ -0,0 +1,69 @@
+
+EVP_PKEY_param_fromdata_init, EVP_PKEY_key_fromdata_init, EVP_PKEY_fromdata,
+EVP_PKEY_param_fromdata_settable, EVP_PKEY_key_fromdata_settable
+- functions to create domain parameters and keys from numbers
```
replace 'numbers' with user data.

levitte replied:
Okie

slontis commented on doc/man3/EVP_PKEY_fromdata.pod:
```diff
@@ -0,0 +1,69 @@
+written to I<*ppkey>.
+
+EVP_PKEY_param_fromdata_settable() and EVP_PKEY_key_fromdata_settable()
+get a constant B<OSSL_PARAM> array that decribes the settable parameters
```
describes

levitte replied:
Okie

slontis commented on test/evp_fromdata_test.c:
```diff
@@ -0,0 +1,87 @@
+
+static int test_fromdata_rsa(void)
+{
+    /* Array indexes */
```
Maybe just move these defines up - The comment looks ugly indented.

levitte replied:
Okie

levitte commented:
Okie, @slontis, I hope I've addressed all your comments

levitte commented:
I rebased, and squashed while I was at it.

@slontis, or anyone really...  final verdict?

slontis commented on CHANGES:
```diff
@@ -9,6 +9,13 @@
 
+  *) Added functionality to create an EVP_PKEY from user data.  This
+     is effectively the same as creating a RSA, DH or DSA object and
+     then assign them to an EVP_PKEY, but directly using algorithm
```
NIT: assigning

levitte replied:
Fixed

levitte committed:
fixup! Add a test for EVP_PKEY_keymake() and EVP_PKEY_make()

levitte committed:
fixup! Add EVP functionality to create domain params and keys by user data

slontis commented on test/evp_fromdata_test.c:
```diff
@@ -0,0 +1,87 @@
+        goto err;
+
+    /*
+     * TODO(3.0) We can't do much more at this point without using internals,
```
I am not sure what you mean here..
Surely the EVP_PKEY_sign operation will work at some point?

levitte replied:
Yes, exactly.  So when that's in place, this test could get some additional code that does something with the key that was created, to verify that the key is functional.

slontis reviewed (approved):
Provided the tests pass, this looks good to me. Good work..

levitte commented:
Merged.

46e2dd05ef1456e3e8fc3d12bd839bae01576c19 Add EVP functionality to create domain params and keys by user data
ab14d2af5386897eba8307c9f3220a6d775c0898 Add a test for EVP_PKEY_keymake() and EVP_PKEY_make()
